pumps/graph_sql.go

```diff
@@ -0,0 +1,174 @@
+package pumps
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/TykTechnologies/tyk-pump/analytics"
+	"github.com/mitchellh/mapstructure"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+	gorm_logger "gorm.io/gorm/logger"
+)
+
+const (
+	GraphSQLPrefix = "GraphSQL-Pump"
+	GraphSQLTable  = "tyk_analytics_graph"
+)
+
+type GraphSQLPump struct {
+	db        *gorm.DB
+	SQLConf   SQLConf
+	tableName string
+	CommonPumpConfig
+}
+
+func (g *GraphSQLPump) GetName() string {
+	return "Graph SQL Pump"
+}
+
+func (g *GraphSQLPump) New() Pump {
+	return &GraphSQLPump{}
+}
+
+func (g *GraphSQLPump) Init(conf interface{}) error {
+	g.log = log.WithField("prefix", GraphSQLPrefix)
+
+	if err := mapstructure.Decode(conf, &g.SQLConf); err != nil {
+		g.log.WithError(err).Error("error decoding conf")
+		return fmt.Errorf("error decoding conf: %w", err)
+	}
+
+	logLevel := gorm_logger.Silent
+
+	switch g.SQLConf.LogLevel {
+	case "debug":
+		logLevel = gorm_logger.Info
+	case "info":
+		logLevel = gorm_logger.Warn
+	case "warning":
+		logLevel = gorm_logger.Error
+	}
+
+	dialect, errDialect := Dialect(&g.SQLConf)
+	if errDialect != nil {
+		g.log.Error(errDialect)
+		return errDialect
+	}
+
+	db, err := gorm.Open(dialect, &gorm.Config{
+		AutoEmbedd:  true,
+		UseJSONTags: true,
+		Logger:      gorm_logger.Default.LogMode(logLevel),
+	})
+	if err != nil {
+		g.log.WithError(err).Error("error opening gorm connection")
+		return err
+	}
+	g.db = db
+
+	if g.SQLConf.BatchSize == 0 {
+		g.SQLConf.BatchSize = SQLDefaultQueryBatchSize
+	}
+
+	g.tableName = GraphSQLTable
+	if name := g.SQLConf.TableName; name != "" {
+		g.tableName = name
+	}
+	if !g.SQLConf.TableSharding {
+		if err := g.db.Table(g.tableName).AutoMigrate(&analytics.GraphRecord{}); err != nil {
+			g.log.WithError(err).Error("error migrating graph analytics table")
+			return err
+		}
+	}
+	g.db = g.db.Table(g.tableName)
+
+	g.log.Debug("pump initialized and table set up")
+	return nil
+}
+
+func (g *GraphSQLPump) WriteData(ctx context.Context, data []interface{}) error {
+	g.log.Debug("Attempting to write ", len(data), " records...")
+
+	var graphRecords []*analytics.GraphRecord
+	for _, r := range data {
```

Comment:
@kofoworola what do you think on moving this conversion logic to a separated function to reduce cognitive complexity? Something like:
```
func (g *GraphSQLPump) getGraphRecords(data []interface{}) ([]*analytics.GraphRecord, error) {
    var graphRecords []*analytics.GraphRecord
    for _, r := range data {
        if r != nil {
            rec, ok := r.(analytics.AnalyticsRecord)
            if !ok || !rec.IsGraphRecord() {
                continue
            }
            gr, err := rec.ToGraphRecord()
            if err != nil {
                g.log.Warnf("error converting 1 record")
                g.log.WithError(err).Debug("error converting record")
                continue
            }
            graphRecords = append(graphRecords, &gr)
        }
    }
    return graphRecords, nil
}
```